### PR TITLE
attestations: Add support for GitHub pull request approval attestation

### PR DESCRIPTION
**Pull request by adityasaky**
~This will change a fair bit with #372.~

Stacked over #489

### CODE REVIEW

File: internal/policy/verify.go
Comment (wlynch) [inline internal/policy/verify.go:150]: Where is the gitObj signature being done? It's not immediately obvious from this code so this looks like we're checking the same thing twice.

File: internal/policy/verify.go
Comment (wlynch) [inline internal/policy/verify.go:159]: What is this tracking and why is it only checked the 2nd time?

File: internal/policy/verify.go
Comment (wlynch) [inline internal/policy/verify.go:177]: Also not clicking what this has to do with file rules. If this is for mergeability, e.g. "the commit has the number of required RSL entries needed to meet the threshold", then this should be renamed since it reads like "RSL entries are needed to meet the threshold"

File: internal/policy/verify.go
Comment (wlynch) [inline internal/policy/verify.go:111]: Why resolve the target ids via the RSL instead of via the refs directly?

Could the RSL potentially be behind the ref value if a new commit was pushed without updating the RSL?

File: internal/policy/verify.go
Comment (wlynch) [inline internal/policy/verify.go:263]: Taking a step back, there's a lot of verification logic here that feels like should be encapsulated elsewhere. Ideally I think this should look like:

What information do you need to be returned from verification in order to make policy decisions about mergability?

File: internal/policy/verify.go
Comment (adityasaky) [inline internal/policy/verify.go:672]: One big question in my mind is that this interface is going from caring strictly about a git sig and dsse sigs to being aware of a specific predicate type. I think we may want to revert this change and handle the special casing of the github approval attestations separately.

**Comment by adityasaky** [comment]
@wlynch I've moved your comments about verify-mergeable over to #455 so this PR can be focused on the github app approval attestation and related functions.

File: internal/policy/verify.go
Comment (adityasaky) [reply internal/policy/verify.go:672]: working on rolling this back

Comment (adityasaky) [reply internal/policy/verify.go]: i've pushed changes on this, i'll clean up tomorrow.

File: internal/cmd/trust/githubappapprovals/githubappapprovals.go
Comment (adityasaky) [inline internal/cmd/trust/githubappapprovals/githubappapprovals.go:37]: this one is a deviation from how other CLI commands work. should this be enable and disable subcommands?

File: internal/tuf/tuf.go
Comment (adityasaky) [reply internal/tuf/tuf.go:89]: I lean towards keeping it vendor specific to start because the attestations we generate may differ. I wonder if the approvers predicate here should be accompanied by a github PR attestation, for example.

**Comment by adityasaky** [comment]
Exploring if we can also track PR numbers here so we can better track dismissals

**Comment by adityasaky** [comment]
This is now tracking reviewID for each approval so that dismissals can be handled. Dismissal tracking isn't perfect but I'd prefer to move ahead with something we can ship in the github app soon and iterate as needed later.

File: internal/policy/root_test.go
Comment (adityasaky) [inline internal/policy/root_test.go:184]: assert.Empty?

File: internal/repository/root_test.go
Comment (adityasaky) [inline internal/repository/root_test.go:328]: assert.Empty

Comment (adityasaky) [reply internal/repository/root_test.go:328]: i wonder why we even have an empty slice, we should probably not initialize it if not needed

File: internal/repository/root_test.go
Comment (adityasaky) [inline internal/repository/root_test.go:350]: this is actually because of state.Verify right? Can we assert a named error?

**Comment by adityasaky** [comment]
@patzielinski thanks for adding tests!

@wlynch I think this is ready for your review.

The predicate now has an approvers list that we use in the policy package to verify. With the refactor:
* more of the github api logic is moved into gittuf rather than handling it in gittuf/github-app
* the predicate has a dismissedApprovers list for those that dismiss a prior approval (this isn't perfect to establishing a timeline of approve - dismiss - approve, but I think it's okay for now because a dismiss is likely because a new change landed meaning the next approve wouldn't be in the same attestation at all)
* we now track the unique github review ID, which is mapped to the <ref>/<from>-<to>; we need this because if a review is dismissed after changes land in either the base branch or the feature branch, we lose the ability to identify exactly which change the review was for when it was first issued

**Comment by adityasaky** [comment]
gotcha, can we open a ticket for this similar to #73? We can probably mock github api responses but we don't have to do it here I reckon.

File: internal/repository/root.go
Comment (adityasaky) [reply internal/repository/root.go:196]: Yeah, it's really more `Set` rather than `Add`.

Comment (adityasaky) [reply internal/repository/root.go:196]: I propose we take a consistency pass for the other APIs as well later

Comment (adityasaky) [reply internal/repository/root.go:196]: Sure

File: internal/repository/attestations.go
Comment (wlynch) [inline internal/repository/attestations.go:400]: If you're defining this multiple times, you're probably better off moving this out into it's own type.

Comment (wlynch) [reply internal/repository/attestations.go:400]: In general, this looks largely similar to AddGitHubPullRequestApprover - can we refactor this to move common parsing/writing to a shared function?

File: internal/attestations/attestations.go
Comment (wlynch) [inline internal/attestations/attestations.go:63]: I'm a bit skeptical of this - I think we need to include remote information of some kind here if we're going to be keying off the pull request id, otherwise it's ambiguous across repos.

I'm also skeptical of using int64 if we ever want to expand this out to other types of presubmit attestations.

File: internal/repository/attestations.go
Comment (wlynch) [inline internal/repository/attestations.go:318]: Rule of thumb: only one-line if it's a very simple struct (usually 1 field)
```suggestion
		stmt := &tmpStatement{
		  Subject: []*ita.ResourceDescriptor{}, 
		  Predicate: &attestations.GitHubPullRequestApprovalAttestation{
		    Approvers: []*tuf.Key{},
		    DismissedApprovers: []*tuf.Key{},
		  }
		}
```

Comment (wlynch) [inline internal/repository/attestations.go:318]: Could this be replaced by `stmt := new(tmpStatement)`?

File: internal/repository/attestations.go
Comment (wlynch) [inline internal/repository/attestations.go:569]: I would add this in a separate change + add docs.

File: internal/policy/root.go
Comment (wlynch) [inline internal/policy/root.go:154]: Is there anything special about this key that's specific to GitHub Apps, or is this a generic "Attestor" role?

Comment (wlynch) [reply internal/policy/root.go:154]: In particular, I want to avoid confusion with https://docs.github.com/en/apps/creating-github-apps/authenticating-with-a-github-app/managing-private-keys-for-github-apps

**Review by wlynch — COMMENTED** [review]
Sorry about the delay here. This is a huge change so it's been hard to find the time to dedicate to sit down and review the whole thing 🙈

File: internal/policy/root.go
Comment (adityasaky) [reply internal/policy/root.go:154]: From the gittuf side, this is special. During the verification flow, we trust the contents of the github app predicate (which says X reviewed this change) only if it's signed by this key specified in the root.

FWIW, (I'll double check), I'm currently reusing the key used by the app to communicate with github to also sign the attestation the app generates, so it's currently the same key. We may want to change this, though, if the app uses an identity + sigstore for example.

https://github.com/gittuf/github-app/pull/1/files#diff-59d92a25c9a864d2527aa8c12116ba798ab65673cf86cd0759b4c1075d01653bR299

File: internal/repository/attestations.go
Comment (adityasaky) [reply internal/repository/attestations.go:318]: gotcha, will fix.

File: internal/attestations/attestations.go
Comment (adityasaky) [reply internal/attestations/attestations.go:63]: just emphasizing, this isn't the PR number but rather the PR ID. I thought that was unique across repositories, but I can fix this to have more data points. TBF, there's also no easy way to look up a PR strictly by ID I think, so having additional information would be good.

int64 is because I'm using the ID field exactly as returned by go-github, so it matches what's used for the field in the API response. Do we expect presubmit attestations to be concentrated in how they're stored in refs/gittuf/attestations? Do you mean if we're supporting gitlab in addition to github?

File: internal/repository/attestations.go
Comment (adityasaky) [reply internal/repository/attestations.go:400]: :eyes:

File: internal/repository/attestations.go
Comment (adityasaky) [reply internal/repository/attestations.go:569]: https://github.com/gittuf/gittuf/pull/489

**Comment by adityasaky** [comment]
@wlynch: I've opened #489 to split off the enterprise github support bits. I've also addressed a bunch of your other comments, including consolidating how we load the approval attestation predicate across the two APIs. I've held off on bigger changes on https://github.com/gittuf/gittuf/pull/439#discussion_r1681856606 and https://github.com/gittuf/gittuf/pull/439#discussion_r1681853263 so you can confirm:
1) we definitely want to store additional information next to the PR ID to track which repository it is
2) how we want to handle the signing key used by the app

I'll test the current version in a test deployment tomorrow.

>This is a huge change so it's been hard to find the time to dedicate to sit down and review the whole thing 🙈

Yeah, this one got bigger than I thought it would 😓

File: internal/attestations/attestations.go
Comment (adityasaky) [reply internal/attestations/attestations.go:63]: Generalized this layer to support multiple code review systems.

File: internal/policy/root.go
Comment (adityasaky) [reply internal/policy/root.go:154]: fixed docs for this

**Comment by adityasaky** [comment]
Note: I force pushed https://github.com/gittuf/gittuf/pull/439/commits/b78206675faafedab841774b2b0453e9d68c31ba because I noticed the dismiss-github-approval CLI command didn't have the `review-iD` parameter set

File: internal/attestations/attestations.go
Comment (adityasaky) [reply internal/attestations/attestations.go:58]: Yes, the predicate types may differ and this supports cases where the same change is approved simultaneously on multiple platforms. I don't know that this makes a lot of sense 99.99% of the time but I wonder if there are strange github+gerrit or so situations out there that we don't want to be backed into a corner on?

File: internal/attestations/attestations.go
Comment (adityasaky) [reply internal/attestations/attestations.go:25]: I was trying to keep the reference authorizations (signed by reviewer) and platform reviews (signed by platform / app on platform) separate, but we could consolidate into one tree. Above, we're recording the platform at the end of the code review path. Should we use `gittuf` for the reference authorizations and consolidate? That also addresses the question above about whether we need to record the provider or not, as I can definitely see gittuf auth and forge approvals used together, even if two different forge provider approvals is rare.

Comment (wlynch) [reply internal/attestations/attestations.go:25]: Yeah the situation I'm hoping to avoid is needed to guess which ref is the right place to place attestations.

The question I'd pose is what value do we get from keeping these separate? In many ways if you're using an app for delegate authorizations that IS the authorization, no different than if a human reviewer directly authorized it. If you're not relying on these directly and instead using it for supplemental information, than that seems similar to if someone approves a review but isn't authorized to be authoritative (e.g. if you don't have OWNERS).

I think my expectation is that these would be stored in the same location rather than needing to search N different ref spaces to find all potential authorizations.
That said, I do think it makes sense to have separate refs for things in the repo vs things being proposed to be added to the repo - we may want to take more liberties to limit history depth to keep the repo history clean for things that aren't necessarily trusted.

Comment (adityasaky) [reply internal/attestations/attestations.go:25]: To clarify a couple things, we're using only one ref for attestations, and currently placing user-signed attestations (reference authorizations) and app-signed-on-behalf-of-scp attestations in two directories. The former is stored as a single blob (<ref>/<from>-<to> is the path in the tree, points to a blob object) while the latter is a directory in which each app can have an entry (<ref>/<from>-<to>/<scp>). I'm a little unclear on the "things being proposed to be added" bit here: do you mean attestations for a change that hasn't been merged yet are stored separately from attestations for changes that have been merged in?

**Review by wlynch — APPROVED** [review]

**Pull request merged**
